### ISSUE TITLE
mypy doesn't complain about non-existent attributes when setting

### ISSUE THREAD

**Issue by samskiter** [pydantic/pydantic#9017]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Related: https://github.com/pydantic/pydantic/issues/245

The above issue is fixed for getting attributes when using the pydantic mypy plugin, but not for setting attributes

### Example Code

```Python3
class blah(pydantic.BaseModel):

    temperature: Optional[int] = pydantic.Field(
        ..., description="The temperature of the zone"
    )
    zone: int = pydantic.Field(..., description="The zone number")


myblah = blah(temperature=1, zone=2)
myblah.sdlkjdshljkdsj = "wekjhe" <<< # does not generate an error

print(myblah.aaa) # <<< Generates an error
```

### Python, Pydantic & OS Version

```Text
python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.6.3
        pydantic-core version: 2.16.3
          pydantic-core build: profile=release pgo=true
                 install path: /Users/samduke/Development/tado-playground/hub_rpi/.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.8 (main, Feb  6 2024, 21:21:21) [Clang 15.0.0 (clang-1500.1.0.2.5)]
                     platform: macOS-14.1-arm64-arm-64bit
             related packages: typing_extensions-4.10.0 mypy-1.8.0
                       commit: unknown
```

**Comment by samskiter** [comment]
Workaround:

Adding an `@no_type_check` decorator to setattr:

```python3
class blah(pydantic.BaseModel):
    @no_type_check
    def __setattr__(self, name: str, value: json.Any) -> None:
        return super().__setattr__(name, value)

    temperature: Optional[int] = pydantic.Field(
        ..., description="The temperature of the zone"
    )
    zone: int = pydantic.Field(..., description="The zone number")

```

**Comment by Viicos** [comment]
Hum, are you certain this is an issue on latest? I believe this was fixed in https://github.com/pydantic/pydantic/pull/8651/.

Those dunder methods are currently defined in a `if not TYPE_CHECKING` block, however it might be interesting to use `no_type_check`, would avoid the extra indentation block

**Comment by sydney-runkle** [comment]
Closing for now, as I believe this is fixed on `main` and will be in 2.7.0b1. Feel free to ping me if you run a test that proves otherwise!
